typescript/src/monitor-printers.ts

```diff
@@ -0,0 +1,138 @@
+    return lastFile && JSON.parse(await readFile(`${statusDir}/${lastFile}`, 'utf-8'));
+}
+
+/**
+ * Compares two lists of cluster statuses.
+ */
+function compareResults(previous: ClusterResponse[], current: ClusterResponse[], dateTime: Date): Comparison {
+    const getOnlineClusterIds = (clusters: ClusterResponse[]): Set<string> => new Set<string>(
+        clusters.filter((cluster) => cluster.is_online).map((cluster) => cluster.cluster_id),
+    );
+    const previousOnlineIds = getOnlineClusterIds(previous);
+    const currentOnlineIds = getOnlineClusterIds(current);
+    return {
+        dateTime: dateTime.toISOString(),
+        appeared: [...currentOnlineIds].filter((id) => !previousOnlineIds.has(id)),
+        gone: [...previousOnlineIds].filter((id) => !currentOnlineIds.has(id)),
+        onlineCount: currentOnlineIds.size,
+        offlineCount: current.length - currentOnlineIds.size,
+    };
+}
+
+/**
+ * Creates a CSV line from the given comparison.
+ */
+function csvLine(comparison: Comparison, separator = ';'): string {
+    const fields: (keyof Comparison)[] = ['dateTime', 'appeared', 'gone', 'onlineCount', 'offlineCount'];
+    return fields.map((field) => comparison[field]).join(separator);
+}
+
+/**
+ * Returns a promise that waits for the given amount of milliseconds, with abort support.
+ */
+function awaitAsync(waitMS: number, abortSignal: AbortSignal): Promise<void> {
+    return new Promise((resolve, reject) => {
+        const onAbort = () => {
+            console.log('Aborting the monitoring...'); // eslint-disable-line no-console
+            resolve();
+        };
+        abortSignal.addEventListener('abort', onAbort);
+        setTimeout(() => {
+            abortSignal.removeEventListener('abort', onAbort);
+            resolve();
+        }, waitMS);
+    });
+}
+
+/**
+ * Gets all clusters from the API periodically, writing the response to a JSON file & the comparisons to a CSV file.
+ */
+async function monitorClusters({
+    comparisonFile = '../cluster-monitoring-logs/cluster-monitoring.csv',
+    waitMS = 60000,
+    statusDir = '../cluster-monitoring-logs',
+}: MonitorClustersOptions = {}): Promise<void> {
+    const abort = new AbortController();
+    process.on('SIGINT', () => abort.abort());
+
+    const demo = new DigitalFactoryDemo();
+    await demo.signIn();
+    print('Sign in completed.\n');
+
+    let previous = await getPreviousClusterStatus(statusDir);
+    const stream = createWriteStream(comparisonFile, { flags: 'a' });
+
+    try {
+        /* eslint-disable no-await-in-loop */
+        while (!abort.signal.aborted) {
+            print('Retrieving clusters...');
+            const clusters = await demo.getClusters();
+            if (!clusters) continue; // eslint-disable-line no-continue
+            const dateTime = new Date();
+            const logFile = `${statusDir}/${dateTime.toISOString().replace(/:/g, '-')}-clusters.json`;
```

Comment:
Good. It is important to make sure we collect the raw data so that we can process and reprocess it later as much as we feel like.